test_util/cluster_api.py

```diff
@@ -201,55 +202,50 @@ def __init__(self, dcos_uri, masters, public_masters, slaves, public_slaves,
 
         # URI must include scheme
         assert dcos_uri.startswith('http')
-        self.scheme = urlparse(dcos_uri).scheme
+        parse_result = urlparse(dcos_uri)
+        self.scheme = parse_result.scheme
+        self.dns_host = parse_result.netloc.split(':')[0]
 
         # Make URI never end with /
         self.dcos_uri = dcos_uri.rstrip('/')
 
-    @staticmethod
-    def _marathon_req_headers():
-        return {'Accept': 'application/json, text/plain, */*'}
+        # Generate request functions
+        for method in self.request_methods:
+            setattr(self, method, self.cluster_request(getattr(requests, method)))
 
-    def _suheader(self, disable_suauth):
-        if not disable_suauth and self.auth_enabled:
-            return self.web_auth_default_user.auth_header
-        return {}
-
-    def get(self, path="", params=None, disable_suauth=False, **kwargs):
-        hdrs = self._suheader(disable_suauth)
-        hdrs.update(kwargs.pop('headers', {}))
-        return requests.get(
-            self.dcos_uri + path, params=params, headers=hdrs, **kwargs)
-
-    def post(self, path="", payload=None, disable_suauth=False, **kwargs):
-        hdrs = self._suheader(disable_suauth)
-        hdrs.update(kwargs.pop('headers', {}))
-        if payload is None:
-            payload = {}
-        return requests.post(self.dcos_uri + path, json=payload, headers=hdrs)
-
-    def delete(self, path="", disable_suauth=False, **kwargs):
-        hdrs = self._suheader(disable_suauth)
-        hdrs.update(kwargs.pop('headers', {}))
-        return requests.delete(self.dcos_uri + path, headers=hdrs, **kwargs)
-
-    def head(self, path="", disable_suauth=False):
-        hdrs = self._suheader(disable_suauth)
-        return requests.head(self.dcos_uri + path, headers=hdrs)
-
-    def node_get(self, node, path="", params=None, disable_suauth=False, **kwargs):
-        """Execute a GET request against the adminrouter on node."""
-        hdrs = self._suheader(disable_suauth)
-        hdrs.update(kwargs.pop('headers', {}))
-
-        if node in self.masters:
-            port = self.adminrouter_master_port[self.scheme]
+    def cluster_request(self, request_fn):
+        """Decorator function for making authorized requests within the cluster
+        """
+        def wrapped_request(path="", node=None, user=self.web_auth_default_user, port=None, **kwargs):
+            """Requests to DC/OS nodes (or dns if unset) using auth headers
+            from default_user (if set)
+            """
+            if node is None:
+                node = self.dns_host
+            if self.auth_enabled and user:
+                hdrs = user.auth_header
+            else:
+                hdrs = {}
+            hdrs.update(kwargs.pop('headers', {}))
+            url = self.get_url(node=node, path=path, port=port)
```

Comment:
This "get url" using urlunparse definitely doesn't feeling like the right thing to use here. Aren't these all just kwargs we're building for the one call?

How about:
```
url = '{scheme}://{node}/{path}'.format(scheme=self.scheme, '{}:{}'.format(node, port) if port else node, path)
logging.info('{}: {}'.format(method, url)
if self.auth_enabled and user:
    kwargs.setdefault('headers', dict()).setdefault('headers', dict()).update(user.auth_header)
requests.request(method, url, **kwargs)
```
(the format string there will not even need the .format() once we get some of the new python fancy which is coming, https://www.python.org/dev/peps/pep-0498/)